### PR TITLE
Add P argument to solve function

### PR DESCRIPTION
**Pull request by juliusgh**
The class `LinearSolver` already supports passing a custom preconditioner matrix `P` to the PETSc solver.
However, this is so far not available through the `solve` function.

This PR adds an argument `P` to the `solve` function that is passed to the `LinearSolver`. At least in my case, it would be convenient to pass a custom preconditioner matrix directly in this way.

### CODE REVIEW

File: firedrake/solving.py [hunk 1]
```diff
@@ -190,6 +198,9 @@ def _la_solve(A, x, b, **kwargs):
     r"""Solve a linear algebra problem.
 
     :arg A: the assembled bilinear form, a :class:`.Matrix`.
+    :arg P: an optional :class:`~.MatrixBase` to construct any
```
Comment (connorjward) [inline firedrake/solving.py:201]: This is in the wrong place in the docstring as it suggests that `P` is the second argument to the function. Should it be documented as a `kwarg`?

File: firedrake/solving.py [hunk 2]
```diff
@@ -52,6 +52,14 @@ def solve(*args, **kwargs):
     :class:`.EquationBC`\s specifying, respectively, the strong boundary conditions
     to apply and PDEs to solve on the boundaries.
     For the format of `solver_parameters` see below.
+    
+    Optionally, an argument `P` of type :class:`~.MatrixBase` can be passed to
+    construct any preconditioner from; if none is supplied `A` is used to
+    construct the preconditioner.
```
Comment (connorjward) [inline firedrake/solving.py:58]:
```suggestion
    Optionally, an argument ``P`` of type :class:`~.MatrixBase` can be passed to
    construct any preconditioner from; if none is supplied ``A`` is used to
    construct the preconditioner.
```

I think. We should check that the docstrings render correctly from the documentation CI artefact.

**Review by connorjward — CHANGES_REQUESTED** [review]
This seems sensible. Thanks!

It would be nice to have a test for this but I think that might be tricky to come up with. We don't even test the `P` argument to `LinearSolver` as things stand.

File: firedrake/solving.py (hunk 1 above)
Comment (juliusgh) [reply firedrake/solving.py]: Yes, I updated it.

**Comment by juliusgh** [comment]
Thanks for the quick reply!

I think a short unit test would indeed be difficult to find. I am using the `P` matrix passing to solve homogenization problems from computational mechanics with a custom preconditioner. FFT-based preconditioners are very effective for these problems, but in addition to `A`, a suitable operator `P` must also be passed. I could provide a documented demo that uses this feature, but maybe that's too specific for firedrake...

**Review by dham — APPROVED** [review]
Thank you!

**Comment by connorjward** [comment]
> Thanks for the quick reply!
> 
> I think a short unit test would indeed be difficult to find. I am using the `P` matrix passing to solve homogenization problems from computational mechanics with a custom preconditioner. FFT-based preconditioners are very effective for these problems, but in addition to `A`, a suitable operator `P` must also be passed. I could provide a documented demo that uses this feature, but maybe that's too specific for firedrake...

To quote @dham re the suggested demo: "yes yes yes please". Demos are always great to have.

**Review by connorjward — DISMISSED** [review]

**Comment by connorjward** [comment]
@juliusgh please could you rebase this code on top of the latest `master`/merge in the latest `master` and push it again? The tests are failing and that should fix them. Then I will merge this. Thanks.

**Comment by juliusgh** [comment]
> @juliusgh please could you rebase this code on top of the latest `master`/merge in the latest `master` and push it again? The tests are failing and that should fix them. Then I will merge this. Thanks.

Ok, I have rebased on top of `master`. I hope that the CI works now.


> To quote @dham re the suggested demo: "yes yes yes please". Demos are always great to have.

Thanks for the feedback! I need some time to finish a well documented demo and would submit a new PR later.

**Review by connorjward — APPROVED** [review]

**Comment by connorjward** [comment]
The rebase appears to have gone a bit wrong (the "Files changed" is now quite large). I have created a fresh PR with just your commits: https://github.com/firedrakeproject/firedrake/pull/3571. I will close this one.

**Comment by juliusgh** [comment]
> The rebase appears to have gone a bit wrong (the "Files changed" is now quite large). I have created a fresh PR with just your commits: #3571. I will close this one.

Oh sorry. Thanks!

**Pull request closed without merging**